React/Project structure.md

```diff
@@ -0,0 +1,361 @@
+
+Example:
+
+```
+.
+├── ...
+└── AlbumsCarousel
+    ├── AlbumsCarousel.tsx
+    └── AlbumsCarousel.utils.ts
+    └── AlbumsCarousel.hooks.ts
+```
+
+**Cluttered component**
+```tsx
+... imports ...
+
+export const AlbumsCarousel = (props) => {
+  const [isPlaying, setIsPlaying] = useState(true);
+  const [highlighted, setHighlighted] = useState(null);
+  const [zoom, setZoom] = useState(1);
+
+  const modals = useModals();
+  const carouselRef = useRef();
+
+  const showLoginModal = () => {
+    modals.open(Modals.Login);
+  };
+
+  const highlightAlbum = (id: number) => {
+    setIsPlaying(false);
+    setHighlighted(id);
+    carouselRef.current.collapseItems();
+  };
+
+  const formatReleaseDate = (startDate: string, endDate: string) => {
+    const start = new Date(startDate);
+    const end = new Date(endDate);
+    const timezonedStart = new Date(start.valueOf() + start.getTimezoneOffset() * 60 * 1000);
+    const timezonedEnd = new Date(end.valueOf() + end.getTimezoneOffset() * 60 * 1000);
+
+    if (isSameDay(timezonedStart, timezonedEnd)) {
+      return `${format(timezonedStart, 'd MMMM yyyy')}`;
+    }
+
+    if (!isSameYear(timezonedStart, timezonedEnd)) {
+      return `${format(timezonedStart, 'd MMMM yyyy')} ${String.fromCharCode(8212)} ${format(
+        timezonedEnd,
+        'd MMMM yyyy',
+      )}`;
+    }
+
+    if (!isSameMonth(timezonedStart, timezonedEnd)) {
+      return `${format(timezonedStart, 'd MMMM')} ${String.fromCharCode(8212)} ${format(
+        timezonedEnd,
+        'd MMMM yyyy',
+      )}`;
+    }
+
+    return `${format(timezonedStart, 'd')} ${String.fromCharCode(8212)} ${format(
+      timezonedEnd,
+      'd MMMM yyyy',
+    )}`;
+  }
+
+  return (
+    //...
+  )
+}
+```
+
+**Cleaned**
+```tsx
+... other imports ...
+import { formatReleaseDate } from './utils';
+
+export const AlbumsCarousel = (props) => {
+  const [isPlaying, setIsPlaying] = useState(true);
+  const [highlighted, setHighlighted] = useState(null);
+  const [zoom, setZoom] = useState(1);
+
+  const modals = useModals();
+  const carouselRef = useRef();
+
+  const showLoginModal = () => {
+    modals.open(Modals.Login);
+  };
+
+  const highlightAlbum = (id: number) => {
+    setIsPlaying(false);
+    setHighlighted(id);
+    carouselRef.current.collapseItems();
+  };
+
+  return (
+    //...
+    // formatReleaseDate used somewhere here
+  )
+}
+```
+
+### Layouts
+
+With Layouts, we generally define overall page structure that will be presented to the user based on a specific auth state or a specific page.
+
+For example, your app could have an Admin dashboard on `/admin` route which has a fixed header and sidebars on both right and left, and scrollable content in the middle (you're most likely familiar with this kind of layout).
+
+You will define your different layouts inside `layouts` folder:
+
+```
+.
+.
+└── components
+    └── layouts
+        ├── admin-layout
```

Comment:
I think folder name should be PascalCase too
```
├── AdminLayout
     └── AdminLayout.tsx
```

Reply:
We could also have some `elements.ts` inside the folder

Reply:
which folder? This is one of the reasons why I want to prefix elements with the "parent" component name, to avoid any standalone elements

Reply:
```
├── admin-layout
     ├── AdminLayout.tsx
     └── AdminLayout.elements.tsx
 ```
 ```
├── AdminLayout
     ├── AdminLayout.tsx
     └── AdminLayout.elements.tsx
 ```
 
 The difference is in the folder name. This is just for consistency.
 - namespace folders `kebab-case`
 - component folders `PascalCase`